add FAQ entry on how to run non-nix executables (#832)
NixOS/nix.dev: source/guides/faq.md

## NixOS

### How to run non-nix executables?

NixOS cannot run dynamically linked executables intended for generic linux environments out of the box.
This is because, by design, it does not have a global library path.

> Solutions include:

Author: This is further from the core issue, in my view. We could support this without being at all FHS-like. `nix-ld` is proof of that.

The core issue is that, unlike distros that shove all their libraries into a shared lookup path, we don't automatically inherit a serviceable library path from the other programs on the system, and would have to explicitly create one solely for non-nix executables, as `nix-ld` does.

Also, I think far more readers will immediately understand that this restriction is somewhat fundamental to the software architecture if we say there is no global library path. Talking about the FHS makes it sound like we just want to be different for not good reason, and don't care about the system being usable.

The way nix packages programs to find their libraries and other dependencies in a separate lookup path for each program should perhaps get an article in "concepts", which could be linked here.

- Use the version packaged in nixpkgs, if there is one.
  You can search available packages at <https://search.nixos.org/packages>.

- Write a nix expression for the program to package it in your own configuration.
  There are multiple approaches:
  - Build from source.
    For an introduction to this, see [](../tutorials/packaging-existing-software).
  - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).
    Do this if building from source isn't feasible.

>   - Wrap the program to run in a normal-seeming container using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments).

Author: I was trying to avoid the word "FHS" in this message. It's jargon that's basically only known in the nixos community, because we need word for what we aren't. The rest of the world doesn't need a word for it. And this message is likely to be many reader's introduction to the fact that nixos is different in this way.

    This is a last resort, but sometimes necessary.
    If the program downloads and runs other executables, for example.

> - Create a library path that only applies to unpackaged programs by using `nix-ld`.

Reviewer: ```suggestion
- Create a library path that only applies to unpackaged programs by using [`nix-ld`](https://github.com/Mic92/nix-ld).
```

>   Note that this does not work for 32-bit executables on `x86_64` machines.

Reviewer: This seems a bit specific, why does it not work? Is it actually due to a fundamental reason or just because of https://github.com/NixOS/nixpkgs/pull/269919?

And this would be best as a note using
```
:::{note}
Text here
:::
```

Author: I only know the `nix-ld` module doesn't try to install it for 32 bit, and the `pkgs.pkgsi686Linux.nix-ld` doesn't build. I didn't look into how difficult it would be to fix the issue.

>   Add this to your `configuration.nix`, rebuild, and relogin to propagate the new environment variables:

Reviewer: ```suggestion
  Add this to your `configuration.nix` and run `nixos-rebuild switch`:
```

  ```nix
  programs.nix-ld.enable = true;
  programs.nix-ld.libraries = with pkgs; [
    # Add any missing dynamic libraries for unpackaged programs
    # here, NOT in environment.systemPackages
  ];

>   ```

Reviewer: ~~~suggestion
  ```
  
  Then log out and back in again.
~~~

A bit harder to miss this way.

Also why is this required at all? Ideally this wouldn't be necessary.

Author: Reworked that so the directions for after editing `configuration.nix` are after the code block.

- Run your program in the normal-seeming container made for the steam package.
  Install `steam-run` and run your program with:

  ```shell-session
  $ steam-run <command>

>   ```

Reviewer: Also we should probably have a package called something like `fhs-run` (or even a `pkgs.fhsRun.withPackages`) that doesn't depend on steam.

## NixOS

### How to run non-nix executables?

NixOS cannot run dynamically linked executables intended for generic linux environments out of the box.

> This is because, by design, it does not have a global library path, nor does it follow the traditional linux standard for where program files are stored on the filesystem.
> 
> :::{note}
> The NixOS community uses the term "FHS" for traditional linux environments, which is short for "[Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html)".
> :::

Reviewer 2: ```suggestion
This is because, by design, it does not have a global library path, nor does it follow the [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html) (FHS).
```

There is nothing special about the Nix community using that abbreviation, and mentioning it should be enough.

Author: We're the only ones who commonly use the term at all, though, abbreviated or not. So it almost certainly needs to be introduced as new jargon to those reading this. I'm not really comfortable using the word here without a prominent introduction to what it means.

I'm a bit on the fence about whether it's prominent enough without the note. It's certainly shorter and more straightforward, which is nice, but not everyone reads linearly, and I want the introduction of the term to catch the eye.

Author: Yeah, alright, I'm good with it. The different coloring of the link is enough, I think.

There are a few ways to resolve this mismatch in environment expectations:

> - Use the version packaged in nixpkgs, if there is one.

Reviewer 2: ```suggestion
- Use the version packaged in Nixpkgs, if there is one.
```

  You can search available packages at <https://search.nixos.org/packages>.

> - Write a nix expression for the program to package it in your own configuration.

Reviewer 2: ```suggestion
- Write a Nix expression for the program to package it for your own configuration.

```

Author: I actually think "in" is better. Users unfamiliar with the nix language often don't realize that they can locate such an expression in or alongside their `configuration.nix`, and assume they need to have a local checkout of nixpkgs in order to have their own packages.

Reviewer 2: Makes sense, but the intent is certainly not evident from just that word. Fixed in a force push.

  There are multiple approaches to this:

>   - Build from source.

Reviewer 2: ```suggestion
  - Build from source.

```

    Many open-source programs are highly flexible at compile time in terms of where their files go.
    For an introduction to this, see [](../tutorials/packaging-existing-software).

>   - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).

Reviewer 2: ```suggestion
  - Set the program's [ELF header](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).

```

    Do this if building from source isn't feasible.

>   - Wrap the program to run in an FHS-like environment using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments).

Reviewer 2: ```suggestion
  - Wrap the program to run in an FHS-like environment using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments),

```

Author: Why make this one end in a comma, when all the other options begin with a sentence saying what the approach is, and then separate sentences with commentary?

>     This is a last resort, but sometimes necessary.
>     If the program downloads and runs other executables, for example.

Reviewer 2: ```suggestion
    This is a last resort, but sometimes necessary, for example if the program downloads and runs other executables.
```

- Create a library path that only applies to unpackaged programs by using [`nix-ld`](https://github.com/Mic92/nix-ld).
  Add this to your `configuration.nix`:

  ```nix
    programs.nix-ld.enable = true;
    programs.nix-ld.libraries = with pkgs; [
      # Add any missing dynamic libraries for unpackaged programs
      # here, NOT in environment.systemPackages
    ];
  ```

  Then run `nixos-rebuild switch`, and log out and back in again to propagate the new environment variables.
  (This is only necessary when enabling `nix-ld`; changes in included libraries take effect immediately on rebuild.)

  :::{note}
  `nix-ld` does not work for 32-bit executables on `x86_64` machines.
  :::

> - Run your program in the FHS-like environment made for the steam package using [`steam-run`](https://nixos.org/manual/nixpkgs/stable/#sec-steam-run):

Reviewer 2: ```suggestion
- Run your program in the FHS-like environment made for the Steam package using [`steam-run`](https://nixos.org/manual/nixpkgs/stable/#sec-steam-run):
```
